src/hardware/mouse/mouseif_ps2_bios.cpp

```diff
@@ -0,0 +1,525 @@
+
+	// 0xff for -1, 0xfe for -2, etc.
+	return static_cast<uint8_t>((tmp >= 0) ? tmp : 0x100 + tmp);
+}
+
+static int16_t get_scaled_movement(const int16_t d)
+{
+	if (!scaling_21)
+		return d;
+
+	switch (d) {
+	case -5: return -9;
+	case -4: return -6;
+	case -3: return -3;
+	case -2: return -1;
+	case -1: return -1;
+	case 1: return 1;
+	case 2: return 1;
+	case 3: return 3;
+	case 4: return 6;
+	case 5: return 9;
+	default: return static_cast<int16_t>(2 * d);
+	}
+}
+
+static void reset_counters()
+{
+	delta_x   = 0.0f;
+	delta_y   = 0.0f;
+	counter_w = 0;
+}
+
+void MOUSEPS2_UpdatePacket()
+{
+	union {
+		uint8_t data = 0x08;
+
+		bit_view<0, 1> left;
+		bit_view<1, 1> right;
+		bit_view<2, 1> middle;
+		bit_view<4, 1> sign_x;
+		bit_view<5, 1> sign_y;
+		bit_view<6, 1> overflow_x;
+		bit_view<7, 1> overflow_y;
+	} mdat;
+
+	mdat.left   = buttons.left;
+	mdat.right  = buttons.right;
+	mdat.middle = buttons.middle;
+
+	auto dx = static_cast<int16_t>(std::round(delta_x));
+	auto dy = static_cast<int16_t>(std::round(delta_y));
+
+	delta_x -= dx;
+	delta_y -= dy;
+
+	dx = get_scaled_movement(dx);
+	dy = get_scaled_movement(static_cast<int16_t>(-dy));
+
+	if (protocol == MouseModelPS2::Explorer) {
+		// There is no overflow for 5-button mouse protocol, see
+		// HT82M30A datasheet
+		dx = std::clamp(dx,
+		                static_cast<int16_t>(-UINT8_MAX),
+		                static_cast<int16_t>(UINT8_MAX));
+		dy = std::clamp(dy,
+		                static_cast<int16_t>(-UINT8_MAX),
+		                static_cast<int16_t>(UINT8_MAX));
+	} else {
+		if ((dx > 0xff) || (dx < -0xff))
+			mdat.overflow_x = 1;
+		if ((dy > 0xff) || (dy < -0xff))
+			mdat.overflow_y = 1;
+	}
+
+	dx %= 0x100;
+	if (dx < 0) {
+		dx          = static_cast<int16_t>(dx + 0x100);
+		mdat.sign_x = 1;
+	}
+
+	dy %= 0x100;
+	if (dy < 0) {
+		dy          = static_cast<int16_t>(dy + 0x100);
+		mdat.sign_y = 1;
+	}
+
+	packet[0] = mdat.data;
+	packet[1] = static_cast<uint8_t>(dx);
+	packet[2] = static_cast<uint8_t>(dy);
+
+	if (protocol == MouseModelPS2::IntelliMouse)
+		packet[3] = get_reset_wheel_8bit();
+	else if (protocol == MouseModelPS2::Explorer) {
+		packet[3] = get_reset_wheel_4bit();
+		if (buttons.extra_1)
+			bit::set(packet[3], b4);
+		if (buttons.extra_2)
+			bit::set(packet[3], b5);
+	} else
+		packet[3] = 0;
+}
+
+static void cmd_set_resolution(const uint8_t new_counts_mm)
+{
+	terminate_unlick_sequence();
+
+	if (new_counts_mm != 1 && new_counts_mm != 2 && new_counts_mm != 4 &&
+	    new_counts_mm != 8)
+		// Invalid parameter, set default
+		counts_mm = 4;
+	else
+		counts_mm = new_counts_mm;
+
+	counts_rate = counts_mm / 4.0f;
+}
+
+static void cmd_set_sample_rate(const uint8_t new_rate_hz)
+{
+	reset_counters();
+
+	if (new_rate_hz != 10 && new_rate_hz != 20 && new_rate_hz != 40 &&
+	    new_rate_hz != 60 && new_rate_hz != 80 && new_rate_hz != 100 &&
+	    new_rate_hz != 200) {
+		// Invalid parameter, set default
```

Comment:
We can't reuse that list, it's different. And when InPort/Bus mouse emulation is introduced, it will have another distinct list. But I have put the allowed values into `std::vector`.